packages/explat-client/src/internal/timing.ts

```diff
@@ -0,0 +1,53 @@
+export const MILLISECONDS_PER_SECOND = 1000;
+
+let lastNow = Date.now();
+/**
+ * Returns the time in miliseconds.
+ * A strictly increasing version of Date.now()
+ */
+export function monotonicNow() {
+	const maybeNow = Date.now();
+
+	lastNow = lastNow < maybeNow ? maybeNow : lastNow + 1;
+
+	return lastNow;
+}
+
+export function createUnresolvingPromise< T >() {
+	return new Promise< T >( ( res ) => {} );
+}
+
+/**
+ * Timeouts a promise. Returns timeoutValue in event of timeout.
+ *
+ * @param promise
+ * @param ms
+ * @param timeoutValue
+ */
+export function timeoutPromise< T, TimeoutValue = null >(
+	promise: Promise< T >,
+	ms: number,
+	timeoutValue: TimeoutValue = null
+): Promise< T | TimeoutValue > {
+	return Promise.race( [
+		promise,
+		new Promise< TimeoutValue >( ( res ) => setTimeout( () => res( timeoutValue ), ms ) ),
+	] );
+}
+
+export function asyncOneAtATime< T >( f: () => Promise< T > ): () => Promise< T > {
```

Comment:
Ah I need to add a comment for this: It wraps an async function so that if it is called multiple times it will just return the same promise - until the promise is fulfilled.

Reply:
The TTL should be doing this if I am understanding this correctly.

Reply:
I mean, there's a brief moment between when this function completes and the result is stored in memory. It'd be good to hold on to the promise for a bit longer to give it time to commit.